internal/exec/stages/disks/luks.go

```diff
@@ -0,0 +1,216 @@
+// Copyright 2020 Red Hat
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The storage stage is responsible for partitioning disks, creating RAID
+// arrays, formatting partitions, writing files, writing systemd units, and
+// writing network units.
+// createRaids creates the raid arrays described in config.Storage.Raid.
+
+package disks
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/coreos/ignition/v2/config/util"
+	"github.com/coreos/ignition/v2/config/v3_2_experimental/types"
+	"github.com/coreos/ignition/v2/internal/distro"
+	execUtil "github.com/coreos/ignition/v2/internal/exec/util"
+)
+
+// https://github.com/latchset/clevis/blob/master/README.md#pin-tang
+type Tang struct {
+	URL        string `json:"url"`
+	Thumbprint string `json:"thp,omitempty"`
+}
+
+// https://github.com/latchset/clevis/blob/master/README.md#pin-shamir-secret-sharing
+type Pin struct {
+	Tpm  bool   `json:"tpm"`
+	Tang []Tang `json:"tang,omitempty"`
+}
+
+func (p Pin) MarshalJSON() ([]byte, error) {
+	if p.Tpm {
+		return json.Marshal(&struct {
+			Tang []Tang   `json:"tang,omitempty"`
+			Tpm  struct{} `json:"tpm2"`
+		}{
+			Tang: p.Tang,
+			Tpm:  struct{}{},
+		})
+	}
+	return json.Marshal(&struct {
+		Tang []Tang `json:"tang"`
+	}{
+		Tang: p.Tang,
+	})
+}
+
+type Clevis struct {
+	Pins      Pin `json:"pins"`
+	Threshold int `json:"t"`
+}
+
+// Initially tested generating keyfiles via dd'ing to a file from /dev/urandom
+// however while cryptsetup had no problem with these keyfiles clevis seemed to
+// die on them while keyfiles generated via openssl rand -hex would work...
+func randHex(length int) (string, error) {
+	bytes := make([]byte, length)
+	if _, err := rand.Read(bytes); err != nil {
+		return "", err
+	}
+	return hex.EncodeToString(bytes), nil
+}
+
+func (s *stage) createLuks(config types.Config) error {
+	if len(config.Storage.Luks) == 0 {
+		return nil
+	}
+
+	for _, luks := range config.Storage.Luks {
+		// TODO: allow Ignition generated KeyFiles for
+		// non-clevis devices that can be persisted.
```

Comment:
Another potential TODO for the future here is creating the LUKS devices in parallel a-la filesystems/raid. It surprisingly takes a lot of time.